### ISSUE TITLE
error when using gzip filehandle with I/O sniffer

### ISSUE THREAD

**Issue by jairideout** [biocore/scikit-bio#836]
When trying to read a gzipped file using scikit-bio's I/O system, the following error is raised while sniffing:

```python
In [1]: from skbio import DistanceMatrix

In [2]: import gzip

In [3]: f = gzip.open('dm.txt.gz', 'rb')

In [4]: DistanceMatrix.read(f)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-0454e36d1310> in <module>()
----> 1 DistanceMatrix.read(f)

/Users/jairideout/dev/scikit-bio/skbio/io/_registry.pyc in read(cls, fp, format, **kwargs)
    698         @classmethod
    699         def read(cls, fp, format=None, **kwargs):
--> 700             return skbio_io_read(fp, into=cls, format=format, **kwargs)
    701
    702         read.__func__.__doc__ = _read_docstring % (

/Users/jairideout/dev/scikit-bio/skbio/io/_registry.pyc in read(fp, format, into, verify, mode, **kwargs)
    599
    600     if format is None:
--> 601         format, fmt_kwargs = sniff(fp, cls=into, mode=mode)
    602         kwargs = _override_kwargs(kwargs, fmt_kwargs, verify)
    603     elif verify:

/Users/jairideout/dev/scikit-bio/skbio/io/_registry.pyc in sniff(fp, cls, mode)
    522             continue
    523         format_sniffer = _sniffers[fmt]
--> 524         is_format, fmt_kwargs = format_sniffer(fp, mode=mode)
    525         if is_format:
    526             possibles.append(fmt)

/Users/jairideout/dev/scikit-bio/skbio/io/_registry.pyc in wrapped_sniffer(fp, mode, **kwargs)
    106                 # true offset of the iterator.
    107                 if hasattr(fh, 'name'):
--> 108                     cfh = open(fh.name, fh.mode)
    109                 else:
    110                     cfh = copy.copy(fh)

TypeError: file() argument 2 must be string, not int
```

If I suppress sniffing by supplying a format and turning off verification, the file is successfully read:

```python
In [1]: from skbio import DistanceMatrix

In [2]: import gzip

In [3]: f = gzip.open('dm.txt.gz', 'rb')

In [4]: DistanceMatrix.read(f, format='lsmat', verify=False)
Out[4]: <skbio.stats.distance._base.DistanceMatrix at 0x10bb13f10>
```

Related to #690.

**Comment by ebolyen** [comment]
Actually it looks like this error is a result of trying to preserve the file mode, which is not needed and will actually be removed with #831

**Comment by jairideout** [comment]
Fixed in #963.
